project5/3700kvstore

```diff
@@ -514,59 +536,62 @@ class Server(LoggingMixin):
+    @dispatch.add('get')
     def handle_put(self, msg):
         """Handles a put request from the client"""
-        if self.state != State.LEADER:
-            # Send fail if leader unknown
-            if self.leader == 'FFFF':
-                self.failed.add('{} {}.put'.format(msg['MID'], msg['src']))
-                self.send_fail(msg)
-            # Else, redirect to the leader
-            else:
-                self.info('{} {}.put redirected to {}'.format(msg['MID'], msg['src'], self.leader))
-                self.send_redirect(msg)
-            return
-
-        # Append entry to local log
-        key = msg['key']
-        value = msg['value']
-        entry = {
-            'command': 'put',
-            'src': msg['src'],
-            'MID': msg['MID'],
-            'key': key,
-            'value': value,
-            'term': self.current_term
-        }
-        self.log.append(entry)
-        self.info('{} {}.put ACCEPTED'.format(msg['MID'], msg['src']))
-        self.broadcast_append_entries()
-
-    @dispatch.add('get')
-    def handle_get(self, msg):
-        """Handles a get request from the client"""
         if self.state == State.LEADER:
-            # Respond with the value if it exists
-            self.send_ok(msg, value=self.datastore.get(msg['key'], ''))
-        elif self.leader == 'FFFF':
-            self.failed.add('{}.get'.format(msg['src']))
-            self.send_fail(msg)
+            for e in self.committed_entries:
+                if e['MID'] == msg['MID']:
+                    self.send_ok(msg, value=self.get(e['key']))
+                    break
+            else:
+                # Append entry to local log
+                key = msg['key']
+                value = msg.get('value', '')
+                entry = {
+                    'command': msg['type'],
+                    'src': msg['src'],
+                    'MID': msg['MID'],
+                    'key': key,
+                    'value': value,
+                    'term': self.current_term
+                }
+                self.log.append(entry)
+                self.info('{} {}.{} ACCEPTED'.format(msg['MID'], msg['src'], msg['type']))
+                self.broadcast_append_entries()
+        # Send fail if leader unknown
         else:
-            self.info('{}.get redirected to {}'.format(msg['src'], self.leader))
-            self.send_redirect(msg)
+            self.fail_or_redirect[msg['MID']] = msg
 
     def apply(self, entries):
         """Commits the entries to the state machine"""
         for e in entries:
             if e['command'] == 'put':
-                self.datastore[e['key']] = e['value']
+                self.put(e['key'], e['value'])
 
     # ------------------------ END OF CLIENT INTERFACE ------------------------
 
     def run(self):
         """Runs the main loop"""
         while True:
             time = current_time()
+            
+            fail_or_redirect_span = time - self.fail_or_redirect_timer
```

Comment:
I think we should pull this into a separate function because `run()` is getting pretty long. It'll also allow you to pull out the `self.fail_or_redirect = {}` line.